CameraAnchor.gd

```diff
@@ -1,34 +1,66 @@
 extends KinematicBody2D
 
 var velocity = Vector2(0, 0)
+var viewport = 0
+var margin_w = 0
+var margin_h = 0
+var move = Vector2(0, 0)
+var mouse_right_click_origin = Vector2(0, 0)
+var mouse_position = Vector2(0, 0)
+
+#Should move one screen width every 4 seconds
+#var move_speed = viewport.size.x / 4
+var move_speed = 16834 # just manually setting it idk
 
 func _ready():
 	pass
 
 func _input(event):
-	var viewport = self.get_viewport()
-	var margin_w = viewport.size.x / 15
-	var margin_h = viewport.size.y / 15
-	var move_x = 0
-	var move_y = 0
-	# Should move one screen width every 4 seconds
-	var move_vel = viewport.size.x / 4
+	viewport = self.get_viewport()
+	margin_w = viewport.size.x / 15
+	margin_h = viewport.size.y / 15
+
 	if event is InputEventMouseMotion:
 		if event.position.x < margin_w:
-			move_x = -1
+			move.x = -1
 		elif event.position.x > viewport.size.x - margin_w:
-			move_x = 1
-		else:
-			move_x = 0
-
+			move.x = 1
+		
 		if event.position.y < margin_h:
-			move_y = -1
+			move.y = -1
 		elif event.position.y > viewport.size.y - margin_h:
-			move_y = 1
-		else:
-			move_y = 0
+			move.y = 1		
+	
+	if Input.is_action_pressed("camera_up"):
+		move.y = -1
+	elif Input.is_action_pressed("camera_down"):
+		move.y = 1
+	
+	if Input.is_action_pressed("camera_left"):
+		move.x = -1
+	elif Input.is_action_pressed("camera_right"):
+		move.x = 1
+	
+	# `normalized() * move_speed` so no fast diagonal movement
+	self.velocity = move.normalized() * move_speed
+	
+	# reset variables
+	move.x = 0
+	move.y = 0
+
+func right_click_movement():
+	if Input.is_action_just_pressed("camera_right_click"):
+		mouse_right_click_origin = get_local_mouse_position()
+	
+	if Input.is_action_pressed("camera_right_click"):
+		mouse_position = get_local_mouse_position()
+		move = mouse_position - mouse_right_click_origin
+		
+		self.velocity = move.normalized() * move_speed * move.length()/256
 		
-		self.velocity = Vector2(move_x * move_vel, move_y * move_vel)
+		move.x = 0
+		move.y = 0
 
```

Comment:
cant all of this be done in the mouse event motion check? by that you have a toggle for the edge scrolling, but if u dont want to do all that ignore it

Reply:
No because `func right_click_movement()` needs to run every frame.

Reply:
told you i did a lot of trial and error :sob:

Reply:
hm. i guess ill leave it there then, can you re-request a review?